## Moive Recommendation

In [2]:
import math
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import re
import seaborn as sns


In [3]:
import os
os.environ["PYSPARK_PYTHON"] = "python3"
# os.environ['PYSPARK_SUBMIT_ARGS'] = '--jars /FileStore/tables/elastic/elasticsearch_spark_20_2_11_7_5_1-cdf5c.jar pyspark-shell'

## Part1: Data ETL and Exploration

In [5]:
from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("moive analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

In [6]:
movies = spark.read.load("/FileStore/tables/movies.csv", format='csv', header = True)
ratings = spark.read.load("/FileStore/tables/ratings.csv", format='csv', header = True)
links = spark.read.load("/FileStore/tables/links.csv", format='csv', header = True)
tags = spark.read.load("/FileStore/tables/tags.csv", format='csv', header = True)

In [7]:
movies.show(5)

+-------+--------------------+--------------------+
movieId| title| genres|
+-------+--------------------+--------------------+
 1| Toy Story (1995)|Adventure|Animati...|
 2| Jumanji (1995)|Adventure|Childre...|
 3|Grumpier Old Men ...| Comedy|Romance|
 4|Waiting to Exhale...|Comedy|Drama|Romance|
 5|Father of the Bri...| Comedy|
+-------+--------------------+--------------------+
only showing top 5 rows

In [8]:
ratings.show(5)

+------+-------+------+----------+
userId|movieId|rating| timestamp|
+------+-------+------+----------+
 1| 307| 3.5|1256677221|
 1| 481| 3.5|1256677456|
 1| 1091| 1.5|1256677471|
 1| 1257| 4.5|1256677460|
 1| 1449| 4.5|1256677264|
+------+-------+------+----------+
only showing top 5 rows

In [9]:
display(ratings.describe())

summary userId movieId rating timestamp count 27753444 27753444 27753444 27753444 mean 141942.01557064414 18487.99983414671 3.5304452124932677 1.1931218549319255E9 stddev 81707.400091494 35102.625247468335 1.066352750231989 2.1604822852234188E8 min 1 1 0.5 1000000065 max 99999 99999 5.0 999999978

In [10]:
# to see if there is null in ratings
ratings.where(ratings['rating'].isNull()).count()

Out[8]: 0

In [11]:
tmp1 = ratings.groupBy("userID").count().toPandas()['count'].min()
tmp2 = ratings.groupBy("movieId").count().toPandas()['count'].min()
print('For the users that rated movies and the movies that were rated:')
print('Minimum number of ratings per user is {}'.format(tmp1))
print('Minimum number of ratings per movie is {}'.format(tmp2))

For the users that rated movies and the movies that were rated:
Minimum number of ratings per user is 1
Minimum number of ratings per movie is 1

In [12]:
tmp1 = sum(ratings.groupBy("movieId").count().toPandas()['count'] == 1)
tmp2 = ratings.select('movieId').distinct().count()
print('{} out of {} movies are rated by only one user'.format(tmp1, tmp2))

10155 out of 53889 movies are rated by only one user

## Spark SQL and OLAP

In [14]:
from pyspark.sql.functions import udf
from pyspark.sql.types import *


In [15]:
movies.createOrReplaceTempView('movies')
ratings.createOrReplaceTempView('ratings')

#### The number of Users

In [17]:
num_user = ratings.select('userId').distinct().count()
num_rating = ratings.select('rating').count()
print('number of users are: {}'.format(num_user))
print('total rating number is: {}'.format(num_rating))

number of users are: 283228
total rating number is: 27753444

#### The number of Movies

In [19]:
num_mv = movies.select('movieId').distinct().count()
print('number of Movies are: {}'.format(num_mv))

number of Movies are: 58098

#### Number of movies rated by users. Also list movies not rated before.

In [21]:
%sql

Select count(distinct movieId) as rated_num
From ratings
Where rating is not null

rated_num 53889

In [22]:
%sql
-- method1 slower

Select distinct movieId, title
From movies
Where movieId not in (Select distinct movieId From ratings Where rating is not null)
Order by movieId

movieId title 100609 Fambul Tok (2011) 100642 Sunday Lovers (1980) 101216 Shadow Boxers (1999) 101229 May the Best Man Win (2009) 101237 9500 Liberty (2009) 101369 Shadow Zone: My Teacher Ate My Homework (1997) 101391 Ascent, The (1994) 101407 Laffghanistan: Comedy Down Range (2009) 101437 Baby Dance, The (1998) 101441 Inspectors 2: A Shred of Evidence, The (2000) 101472 Brussels Business, The (2012) 102082 Always in My Heart (1942) 102135 Sadda Haq (2013) 102150 Angel from Texas, An (1940) 102192 Successful Calamity, A (1932) 102475 Anne B. Real (2003) 103968 And Baby Makes Three (Baby Is Here) (1949) 103970 And So They Were Married (Bless Their Hearts) (1936) 104024 Annie O (1996) 104026 Another Face (Two Faces) (1935) 104095 Attack, The (1996) 104261 80 Steps to Jonah (1969) 104417 Cry in the Night, A (1992) 104656 Across the Sierras (1941) 105111 Bandit: Bandit's Silver Angel (Smokey and the Bandit 7) (Bandit: Bandit and the Silver Angel) (1994) 105279 Adventures of Kitty O'Day (Kitty O'Day Comes Through) (1945) 105542 Another Harvest Moon (2010) 105589 Bachelor Bait (1934) 105792 Stranger in Town, A (1943) 105796 Back in the Saddle (Back in the Saddle Again) (1941) 106028 Thrill of Brazil, The (1946) 106034 Don't Play Us Cheap (1973) 106037 Dreamer (1979) 106282 180 Segundos (2012) 106284 Back Pay (1930) 106304 Hit! (1973) 106419 Aggie Appleby Maker of Men (1933) 107289 American Raspberry (Prime Time) (Funny America) (1977) 107291 Annapolis Salute (1937) 107432 Green Hell (1940) 107442 Two Against Time (2002) 107774 When the North Wind Blows (1974) 107776 Across the Great Divide (1976) 107838 Attorney for the Defense (1932) 108046 Soldier's Plaything, A (1930) 108125 Allotment Wives (1945) 108304 Another Dawn (1937) 108374 Arizona (Men Are Like That) (Virtuous Wife, The) (1931) 108432 Guru, The (1969) 108637 Al Jennings of Oklahoma (1951) 108643 America (2011) 108679 Bad Bascomb (1946) 108707 Nero Wolfe (1979) 108717 Moving Target (2000) 108743 Bannerline (1951) 108816 Sleep Room, The (1998) 108844 BBOY for LIFE (2012) 108846 Before and After (1979) 108918 Beloved Infidel (1959) 108922 Best Friends (1975) 108995 Between Two Women (Surrounded by Women) (1937) 109078 Beyond the Call (1996) 109233 Beyond the Purple Hills (1950) 109271 Big Leaguer (1953) 109512 Anna Lucasta (1949) 109645 Behind Office Doors (Private Secretary) (1931) 109647 Bengazi (1955) 109706 Bittersweet Love (1976) 109745 Black Hills (1947) 109837 Blackjack Ketchum, Desperado (1956) 109917 Blond Cheat (Blonde Cheat) (1938) 109979 Blonde Inspiration (1941) 110065 Stunts (1977) 110088 Blood on the Arrow (1964) 110151 20 Mule Team (1940) 110153 Challenge for Robin Hood, A (1967) 110155 Man Called Adam, A (1966) 110157 April Showers (1948) 110161 Battle of Rogue River (1954) 110212 Bloodfist VIII: Trained to Kill (Hard Way Out) (1996) 110248 Blue Canadian Rockies (1952) 110250 Blue Montana Skies (1939) 110354 Trailer Park Jesus (2012) 110370 Barabbas (2012) 110384 Last Word, The (1979) 110399 Boots and Saddles (1937) 110401 Boots Malone (1952) 110476 Border Treasure (1950) 110488 Bled (2009) 110493 Boston's Finest (2010) 110495 Boundaries of the Heart (1988) 110517 Old Drac (Vampira) (1974) 110737 Old Boyfriends (1979) 110820 Liberation of L.B. Jones, The (1970) 110822 Super Fly T.N.T. (1973) 110856 Broadway to Hollywood (1933) 110916 Study in Scarlet, A (1933) 110919 All Hands on Deck (1961) 110923 Before Dawn (1933) 111048 Good Day to Be Black & Sexy, A (2008) 111150 Always a Bride (1953) 111307 Visit to a Chief's Son (1974) 111325 Neptune Factor, The (1973) 111358 Small Town in Texas, A (1976) 111523 50 Kisses (2014) 111525 Kiss in the Dark, A (1949) 111527 A Xiu Luo (Saga of the Phoenix) (1990) 111560 Affectionately Yours (1941) 111597 Heroic Duo (Shuang xiong) (2003) 111599 Lady Ice (1973) 111667 Toward the Unknown (1956) 111678 Army Surgeon (1942) 111682 Baby Geniuses and the Mystery of the Crown Jewels (Baby Geniuses: Baby Squad Investigators) (Baby Ge

In [23]:
%sql
-- method2 faster

Select distinct m.movieId, title
From movies m
Left Join ratings r on m.movieId=r.movieId
Where r.rating is null
Order by movieId

movieId title 100609 Fambul Tok (2011) 100642 Sunday Lovers (1980) 101216 Shadow Boxers (1999) 101229 May the Best Man Win (2009) 101237 9500 Liberty (2009) 101369 Shadow Zone: My Teacher Ate My Homework (1997) 101391 Ascent, The (1994) 101407 Laffghanistan: Comedy Down Range (2009) 101437 Baby Dance, The (1998) 101441 Inspectors 2: A Shred of Evidence, The (2000) 101472 Brussels Business, The (2012) 102082 Always in My Heart (1942) 102135 Sadda Haq (2013) 102150 Angel from Texas, An (1940) 102192 Successful Calamity, A (1932) 102475 Anne B. Real (2003) 103968 And Baby Makes Three (Baby Is Here) (1949) 103970 And So They Were Married (Bless Their Hearts) (1936) 104024 Annie O (1996) 104026 Another Face (Two Faces) (1935) 104095 Attack, The (1996) 104261 80 Steps to Jonah (1969) 104417 Cry in the Night, A (1992) 104656 Across the Sierras (1941) 105111 Bandit: Bandit's Silver Angel (Smokey and the Bandit 7) (Bandit: Bandit and the Silver Angel) (1994) 105279 Adventures of Kitty O'Day (Kitty O'Day Comes Through) (1945) 105542 Another Harvest Moon (2010) 105589 Bachelor Bait (1934) 105792 Stranger in Town, A (1943) 105796 Back in the Saddle (Back in the Saddle Again) (1941) 106028 Thrill of Brazil, The (1946) 106034 Don't Play Us Cheap (1973) 106037 Dreamer (1979) 106282 180 Segundos (2012) 106284 Back Pay (1930) 106304 Hit! (1973) 106419 Aggie Appleby Maker of Men (1933) 107289 American Raspberry (Prime Time) (Funny America) (1977) 107291 Annapolis Salute (1937) 107432 Green Hell (1940) 107442 Two Against Time (2002) 107774 When the North Wind Blows (1974) 107776 Across the Great Divide (1976) 107838 Attorney for the Defense (1932) 108046 Soldier's Plaything, A (1930) 108125 Allotment Wives (1945) 108304 Another Dawn (1937) 108374 Arizona (Men Are Like That) (Virtuous Wife, The) (1931) 108432 Guru, The (1969) 108637 Al Jennings of Oklahoma (1951) 108643 America (2011) 108679 Bad Bascomb (1946) 108707 Nero Wolfe (1979) 108717 Moving Target (2000) 108743 Bannerline (1951) 108816 Sleep Room, The (1998) 108844 BBOY for LIFE (2012) 108846 Before and After (1979) 108918 Beloved Infidel (1959) 108922 Best Friends (1975) 108995 Between Two Women (Surrounded by Women) (1937) 109078 Beyond the Call (1996) 109233 Beyond the Purple Hills (1950) 109271 Big Leaguer (1953) 109512 Anna Lucasta (1949) 109645 Behind Office Doors (Private Secretary) (1931) 109647 Bengazi (1955) 109706 Bittersweet Love (1976) 109745 Black Hills (1947) 109837 Blackjack Ketchum, Desperado (1956) 109917 Blond Cheat (Blonde Cheat) (1938) 109979 Blonde Inspiration (1941) 110065 Stunts (1977) 110088 Blood on the Arrow (1964) 110151 20 Mule Team (1940) 110153 Challenge for Robin Hood, A (1967) 110155 Man Called Adam, A (1966) 110157 April Showers (1948) 110161 Battle of Rogue River (1954) 110212 Bloodfist VIII: Trained to Kill (Hard Way Out) (1996) 110248 Blue Canadian Rockies (1952) 110250 Blue Montana Skies (1939) 110354 Trailer Park Jesus (2012) 110370 Barabbas (2012) 110384 Last Word, The (1979) 110399 Boots and Saddles (1937) 110401 Boots Malone (1952) 110476 Border Treasure (1950) 110488 Bled (2009) 110493 Boston's Finest (2010) 110495 Boundaries of the Heart (1988) 110517 Old Drac (Vampira) (1974) 110737 Old Boyfriends (1979) 110820 Liberation of L.B. Jones, The (1970) 110822 Super Fly T.N.T. (1973) 110856 Broadway to Hollywood (1933) 110916 Study in Scarlet, A (1933) 110919 All Hands on Deck (1961) 110923 Before Dawn (1933) 111048 Good Day to Be Black & Sexy, A (2008) 111150 Always a Bride (1953) 111307 Visit to a Chief's Son (1974) 111325 Neptune Factor, The (1973) 111358 Small Town in Texas, A (1976) 111523 50 Kisses (2014) 111525 Kiss in the Dark, A (1949) 111527 A Xiu Luo (Saga of the Phoenix) (1990) 111560 Affectionately Yours (1941) 111597 Heroic Duo (Shuang xiong) (2003) 111599 Lady Ice (1973) 111667 Toward the Unknown (1956) 111678 Army Surgeon (1942) 111682 Baby Geniuses and the Mystery of the Crown Jewels (Baby Geniuses: Baby Squad Investigators) (Baby Ge

#### conclusion 1

In this dataset, we have 58098 movies, only 53889 of them are rated. There are 4209 of them are not rated by users.

#### List Movie Genres

In [26]:
%sql

Select distinct explode(split(genres,'\\|'))
From movies

col Crime Romance Thriller Adventure Drama War Documentary Fantasy Mystery Musical Animation Film-Noir (no genres listed) IMAX Horror Western Comedy Children Action Sci-Fi

### Number of movies and average rating for each category

In [28]:
categories = spark.sql(
                   '''Select cte.category, count(*) as rate_num, avg(rating) as avg_rate
                      From movies m
                      Join
                        (Select movieId, title, explode(split(genres,'\\\\|')) as category
                        From movies) cte
                      on m.movieId=cte.movieId
                      Join ratings r on m.movieId=r.movieId
                      Group by cte.category
                      Order by 2 Desc''')

In [29]:
categories.show(5)

+---------+--------+------------------+
 category|rate_num| avg_rate|
+---------+--------+------------------+
 Drama|12191048|3.6763707271105814|
 Comedy| 9999184| 3.419786704595095|
 Action| 8214956|3.4625655329133838|
 Thriller| 7489628|3.5188010672893233|
Adventure| 6452309|3.5132836167641694|
+---------+--------+------------------+
only showing top 5 rows

In [30]:
fig, (ax1, ax2) = plt.subplots(figsize=(10,6), ncols=2, constrained_layout=True)
categories_p = categories.toPandas()

xticks = [cate[0] for cate in categories.select("category").collect()]

ax1.plot('category','avg_rate',data=categories_p)
# plt.plot('iter_num','validation_mse',data=categories_p)
ax1.set_xlabel("categories")
ax1.set_xticklabels(xticks,rotation=90)
ax1.set_ylabel("average rating")
ax1.set_title("average rating for each categories", y=1.08)
ax1.legend()

ax2.plot('category','rate_num',data=categories_p)
ax2.set_xlabel("categories")
ax2.set_xticklabels(xticks,rotation=90)
ax2.set_ylabel("rating number")
ax2.set_title("rating number for each categories", y=1.08)
ax2.legend()

display(fig.show())

#### Conclusion 2

We have 20 categories of movies in the dataset. **Drama**, **Comedy** and **Action** are top 3 categories received most rating (around 10 million) by users. **Drama**, **Crime**, **Mystery**, **War** and **Film-Noir** are categories rated by relatively higher score. The average rating scores for all these categories are between 3.3 and 3.9 (rating in 0-5 range).

### Number of each rating score

In [33]:
%sql

Select rating, count(*) as num
From ratings
Group by rating
Order by 1

rating num 0.5 442388 1.0 886233 1.5 441354 2.0 1850627 2.5 1373419 3.0 5515668 3.5 3404360 4.0 7394710 4.5 2373550 5.0 4071135

### Top 1000 movies received most rating

In [35]:
top_rating_num = spark.sql(
                '''Select r.movieId, m.title, sum(case when rating is not null then 1 else 0 end) as rate_num
                From ratings r
                Join movies m on r.movieId=m.movieId
                Group by 1,2
                Order by 3 desc
                Limit 1000''')


In [36]:
def get_movie_year(title):
  matches = re.search("\(\d{4}\)", title)
  if matches:
    group = matches.group()
    year = group[1:-1]
#     start = matches.start()
    title = title[:matches.start()-1]
    return (title, year)
  else:
    return (title, None)
    
get_year = udf(get_movie_year,\
               StructType([StructField("title", StringType(), True),\
                           StructField("release_year", StringType(), True)]))


In [37]:
top_rating_num_movies = top_rating_num.select(
                                "movieId", get_year("title").title.alias("title"),\
                                get_year("title").release_year.alias("release_year"),\
                                "rate_num")

display(top_rating_num_movies)

top_rating_num_movies.createOrReplaceTempView('top_rate_num_movies')

movieId title release_year rate_num 318 Shawshank Redemption, The 1994 97999 356 Forrest Gump 1994 97040 296 Pulp Fiction 1994 92406 593 Silence of the Lambs, The 1991 87899 2571 Matrix, The 1999 84545 260 Star Wars: Episode IV - A New Hope 1977 81815 480 Jurassic Park 1993 76451 527 Schindler's List 1993 71516 110 Braveheart 1995 68803 1 Toy Story 1995 68469 1210 Star Wars: Episode VI - Return of the Jedi 1983 66023 1196 Star Wars: Episode V - The Empire Strikes Back 1980 65822 2959 Fight Club 1999 65678 589 Terminator 2: Judgment Day 1991 64258 1198 Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) 1981 63505 50 Usual Suspects, The 1995 62180 4993 Lord of the Rings: The Fellowship of the Ring, The 2001 61883 858 Godfather, The 1972 60904 2858 American Beauty 1999 60820 780 Independence Day (a.k.a. ID4) 1996 58949 150 Apollo 13 1995 58665 457 Fugitive, The 1993 58031 1270 Back to the Future 1985 57492 7153 Lord of the Rings: The Return of the King, The 2003 57378 5952 Lord of the Rings: The Two Towers, The 2002 56696 47 Seven (a.k.a. Se7en) 1995 55240 32 Twelve Monkeys (a.k.a. 12 Monkeys) 1995 54846 608 Fargo 1996 54500 592 Batman 1989 54448 2028 Saving Private Ryan 1998 54027 2762 Sixth Sense, The 1999 52270 590 Dances with Wolves 1990 51877 588 Aladdin 1992 51827 380 True Lies 1994 50864 3578 Gladiator 2000 48666 364 Lion King, The 1994 47305 4306 Shrek 2001 46826 377 Speed 1994 46475 344 Ace Ventura: Pet Detective 1994 45608 648 Mission: Impossible 1996 45064 1721 Titanic 1997 44787 58559 Dark Knight, The 2008 44741 1580 Men in Black (a.k.a. MIB) 1997 44287 4226 Memento 2000 43739 1197 Princess Bride, The 1987 42878 1704 Good Will Hunting 1997 42769 595 Beauty and the Beast 1991 42400 1240 Terminator, The 1984 42254 1193 One Flew Over the Cuckoo's Nest 1975 42181 1291 Indiana Jones and the Last Crusade 1989 42043 79132 Inception 2010 41475 165 Die Hard: With a Vengeance 1995 41244 1136 Monty Python and the Holy Grail 1975 40866 1265 Groundhog Day 1993 40836 1097 E.T. the Extra-Terrestrial 1982 40328 1036 Die Hard 1988 40284 6539 Pirates of the Caribbean: The Curse of the Black Pearl 2003 39674 541 Blade Runner 1982 39441 1214 Alien 1979 39282 1221 Godfather: Part II, The 1974 38875 367 Mask, The 1994 38699 736 Twister 1996 38662 153 Batman Forever 1995 38647 597 Pretty Woman 1990 38504 500 Mrs. Doubtfire 1993 38484 733 Rock, The 1996 37936 231 Dumb & Dumber (Dumb and Dumber) 1994 37823 1089 Reservoir Dogs 1992 37331 2628 Star Wars: Episode I - The Phantom Menace 1999 37179 4973 Amelie (Fabuleux destin d'Amélie Poulain, Le) 2001 37167 4886 Monsters, Inc. 2001 37112 6377 Finding Nemo 2003 37000 316 Stargate 1994 36996 293 Léon: The Professional (a.k.a. The Professional) (Léon) 1994 36504 1527 Fifth Element, The 1997 36269 3793 X-Men 2000 36030 1682 Truman Show, The 1998 36016 1213 Goodfellas 1990 35934 34 Babe 1995 35903 6874 Kill Bill: Vol. 1 2003 35659 2716 Ghostbusters (a.k.a. Ghost Busters) 1984 35343 7361 Eternal Sunshine of the Spotless Mind 2004 35064 1200 Aliens 1986 34572 349 Clear and Present Danger 1994 34547 2329 American History X 1998 34110 4995 Beautiful Mind, A 2001 33936 4963 Ocean's Eleven 2001 33429 1073 Willy Wonka & the Chocolate Factory 1971 33411 2997 Being John Malkovich 1999 33390 5445 Minority Report 2002 33163 3147 Green Mile, The 1999 33140 10 GoldenEye 1995 33086 111 Taxi Driver 1976 32885 586 Home Alone 1990 32738 1961 Rain Man 1988 32726 1617 L.A. Confidential 1997 32553 1206 Clockwork Orange, A 1971 32436 587 Ghost 1990 32210 1258 Shining, The 1980 32129 5349 Spider-Man 2002 32068 33794 Batman Begins 2005 32027 8961 Incredibles, The 2004 31857 539 Sleepless in Seattle 1993 31775 778 Trainspotting 1996 31545 924 2001: A Space Odyssey 1968 31418 208 Waterworld 1995 31310 253 Interview with the Vampire: The Vampire Chronicles 1994 31302 3996 Crouching Tiger, Hidden Dragon (Wo hu cang long) 2000 31224 912 Casablanca 1942 31095 5418 Bourne Identity, The 2002 30822 357 Four Wed

In [38]:
%sql

Select release_year, count(*) as num_by_year
From top_rate_num_movies
Group by release_year
Order by 2 Desc

release_year num_by_year 1995 68 1996 56 1999 51 1994 50 2000 48 1998 47 1997 41 2004 39 2001 38 1993 38 2002 34 1992 26 2003 24 1989 23 2007 22 2005 22 2006 21 1984 18 1991 18 1990 18 1986 16 1987 16 2008 15 2010 15 2009 14 1988 13 1982 13 2012 12 2011 11 2014 10 1981 10 1983 9 1979 9 1980 9 1985 9 2013 8 2015 7 1964 5 1967 5 1968 5 1978 5 1975 5 1959 4 1962 4 1974 4 1973 4 1951 4 1971 4 1954 3 1941 3 1976 3 1940 3 1970 3 2016 3 1977 3 1963 3 1961 3 1942 2 1939 2 1957 2 1953 2 1969 2 1950 2 1955 2 1972 2 1960 1 1937 1 1966 1 1958 1 1952 1 1946 1 1933 1 1965 1 1949 1 1944 1

In [39]:
%sql

Select (case when release_year<1970 then 'before 1970' when release_year between 1970 and 1979 then '1970s' when release_year between 1980 and 1989 then '1980s' when release_year between 1990 and 1999 then '1990s' when release_year between 2000 and 2009 then '2000s' else 'after 2010' end) as decade, count(*) as num_by_decade
From top_rate_num_movies
Group by 1
Order by 2 Desc

decade num_by_decade 1990s 413 2000s 277 1980s 136 before 1970 66 after 2010 66 1970s 42

### Top 1000 movies with highest average rating (received over 1000 ratings)

In [41]:

top_avg_rating_movies = spark.sql(
                  '''Select r.movieId, m.title, round(avg(rating), 2) as avg_rate, sum(case when r.rating is not null then 1 else 0 end) as rate_num
                  From ratings r
                  Join movies m on r.movieId=m.movieId
                  Group by 1,2
                  Having sum(case when r.rating is not null then 1 else 0 end)>=1000
                  Order by 3 desc
                  Limit 1000''')

In [42]:
top_avg_rating_movies = top_avg_rating_movies.select(
                                "movieId", get_year("title").title.alias("title"),\
                                get_year("title").release_year.alias("release_year"),\
                                "avg_rate", "rate_num")

top_avg_rating_movies.createOrReplaceTempView('top_avg_rate_movies')

display(top_avg_rating_movies)

movieId title release_year avg_rate rate_num 159817 Planet Earth 2006 4.46 1384 318 Shawshank Redemption, The 1994 4.42 97999 174053 Black Mirror: White Christmas 2014 4.35 1074 858 Godfather, The 1972 4.33 60904 50 Usual Suspects, The 1995 4.29 62180 527 Schindler's List 1993 4.26 71516 1221 Godfather: Part II, The 1974 4.26 38875 2019 Seven Samurai (Shichinin no samurai) 1954 4.25 14578 1203 12 Angry Men 1957 4.24 17931 2959 Fight Club 1999 4.23 65678 904 Rear Window 1954 4.23 22264 1193 One Flew Over the Cuckoo's Nest 1975 4.22 42181 912 Casablanca 1942 4.21 31095 750 Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb 1964 4.21 29484 5618 Spirited Away (Sen to Chihiro no kamikakushi) 2001 4.21 23227 44555 Lives of Others, The (Das leben der Anderen) 2006 4.2 9670 1212 Third Man, The 1949 4.2 7980 3435 Double Indemnity 1944 4.2 5824 1178 Paths of Glory 1957 4.2 4508 908 North by Northwest 1959 4.2 19613 922 Sunset Blvd. (a.k.a. Sunset Boulevard) 1950 4.2 8264 3030 Yojimbo 1961 4.18 4328 6016 City of God (Cidade de Deus) 2002 4.18 21558 1213 Goodfellas 1990 4.18 35934 926 All About Eve 1950 4.17 5632 296 Pulp Fiction 1994 4.17 92406 58559 Dark Knight, The 2008 4.17 44741 1284 Big Sleep, The 1946 4.16 6470 1260 M 1931 4.16 5056 930 Notorious 1946 4.16 5618 79132 Inception 2010 4.16 41475 2324 Life Is Beautiful (La Vita è bella) 1997 4.16 26995 950 Thin Man, The 1934 4.15 3704 593 Silence of the Lambs, The 1991 4.15 87899 1217 Ran 1985 4.15 5705 1252 Chinatown 1974 4.15 19084 1136 Monty Python and the Holy Grail 1975 4.15 40866 2571 Matrix, The 1999 4.15 84545 4226 Memento 2000 4.15 43739 5971 My Neighbor Totoro (Tonari no Totoro) 1988 4.15 9767 1248 Touch of Evil 1958 4.15 5333 1207 To Kill a Mockingbird 1962 4.15 17988 1234 Sting, The 1973 4.14 17906 160718 Piper 2016 4.14 1253 2329 American History X 1998 4.14 34110 1945 On the Waterfront 1954 4.13 6593 913 Maltese Falcon, The 1941 4.13 14585 1196 Star Wars: Episode V - The Empire Strikes Back 1980 4.13 65822 1147 When We Were Kings 1996 4.13 4084 5291 Rashomon (Rashômon) 1950 4.13 5046 3134 Grand Illusion (La grande illusion) 1937 4.13 1829 112552 Whiplash 2014 4.13 10907 1148 Wallace & Gromit: The Wrong Trousers 1993 4.13 17337 905 It Happened One Night 1934 4.13 4750 2203 Shadow of a Doubt 1943 4.13 2548 92259 Intouchables 2011 4.13 13573 2186 Strangers on a Train 1951 4.12 6137 1172 Cinema Paradiso (Nuovo cinema Paradiso) 1989 4.12 11452 86504 Voices from the List 2004 4.12 1800 260 Star Wars: Episode IV - A New Hope 1977 4.12 81815 3000 Princess Mononoke (Mononoke-hime) 1997 4.12 13865 2858 American Beauty 1999 4.12 60820 48516 Departed, The 2006 4.12 26162 541 Blade Runner 1982 4.12 39441 2920 Children of Paradise (Les enfants du paradis) 1945 4.12 1166 903 Vertigo 1958 4.12 17926 1197 Princess Bride, The 1987 4.12 42878 1201 Good, the Bad and the Ugly, The (Buono, il brutto, il cattivo, Il) 1966 4.12 19731 1280 Raise the Red Lantern (Da hong deng long gao gao gua) 1991 4.12 3609 1254 Treasure of the Sierra Madre, The 1948 4.12 5932 4973 Amelie (Fabuleux destin d'Amélie Poulain, Le) 2001 4.12 37167 1198 Raiders of the Lost Ark (Indiana Jones and the Raiders of the Lost Ark) 1981 4.12 63505 163134 Your Name. 2016 4.12 1205 6669 Ikiru 1952 4.11 1551 898 Philadelphia Story, The 1940 4.11 7828 2731 400 Blows, The (Les quatre cents coups) 1959 4.11 3524 3089 Bicycle Thieves (a.k.a. The Bicycle Thief) (a.k.a. The Bicycle Thieves) (Ladri di biciclette) 1948 4.11 4145 1209 Once Upon a Time in the West (C'era una volta il West) 1968 4.11 5952 2351 Nights of Cabiria (Notti di Cabiria, Le) 1957 4.11 1393 1208 Apocalypse Now 1979 4.11 28986 608 Fargo 1996 4.11 54500 1233 Boot, Das (Boat, The) 1981 4.11 15936 1204 Lawrence of Arabia 1962 4.11 15415 1949 Man for All Seasons, A 1966 4.11 2340 745 Wallace & Gromit: A Close Shave 1995 4.11 13866 2905 Sanjuro (Tsubaki Sanjûrô) 1962 4.11 1499 3307 City Lights 1931 4.11 3363 7153 Lord of the Rings: The Return of the King

In [43]:
%sql

Select (case when release_year<1970 then 'before 1970' when release_year between 1970 and 1979 then '1970s' when release_year between 1980 and 1989 then '1980s' when release_year between 1990 and 1999 then '1990s' when release_year between 2000 and 2009 then '2000s' else 'after 2010' end) as decade, count(*) as num_by_decade, round(sum(avg_rate*rate_num)/sum(rate_num), 2) as avg_score_by_decade
From top_avg_rate_movies
Group by 1
Order by 2 Desc, 3 Desc

decade num_by_decade avg_score_by_decade 2000s 234 3.92 before 1970 218 4.02 1990s 201 4.0 after 2010 138 3.91 1980s 125 3.96 1970s 84 4.04

#### Conclusion 3

Rating scores are mainly distributed between 3.0 and 5.0.   
The top 3 most popular movies are 'Shawshank Redemption', 'Forrest Gump' and 'Pulp Fiction' which receive more than 90,000 rating.   
1990s has 413 of top 1000 popular movies (received most rating in this dataset).   
'Planet Earth', 'Shawshank Redemption' and 'Black Mirror: White Christmas' are the top 3 movies with the highest average rating score around 4.4. While 'Shawshank Redemption', 'The Godfather' and 'The Usual Suspects' are movies both received high number of rating and average rating score. These movies should be recommended to users in general.

## Part2: Spark ALS based approach for training model

In [46]:
from pyspark.ml.recommendation import ALS, ALSModel
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.tuning import ParamGridBuilder, CrossValidator
from pyspark.sql import Row
from pyspark.sql.functions import arrays_zip, col, explode

In [47]:
ratings = ratings['userId','movieId','rating']

ratings = ratings.withColumn("userId", ratings["userId"].cast(LongType()))
ratings = ratings.withColumn("movieId", ratings["movieId"].cast(LongType()))
ratings = ratings.withColumn("rating", ratings["rating"].cast(FloatType()))


#### Now split the data into training/validation/testing sets using a 6/2/2 ratio. Also split it by a 3/1 ratio to use k-fold cross validation.

In [49]:
train, test = ratings.randomSplit([0.75,0.25],seed = 42)
training, validation, testing = ratings.randomSplit([0.6, 0.2, 0.2],seed = 40)

In [50]:
print('train size:', train.count())
print('test size:', test.count())

train size: 20817599
test size: 6935845

### ALS Model Selection and Evaluation

#### Method 1: use cross-validation

In [53]:
ranks = [6, 8, 10]
reg_params = [0.05, 0.1, 0.2]
als = ALS(maxIter=10, userCol="userId", itemCol="movieId", ratingCol="rating",
          coldStartStrategy="drop")

paramGrid = ParamGridBuilder() \
          .addGrid(ALS.rank, ranks) \
          .addGrid(ALS.regParam, reg_params) \
          .build()

mse_evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")

CV = CrossValidator(estimator=als,
                   estimatorParamMaps=paramGrid,
                   evaluator=mse_evaluator,
                   numFolds=3)

cvModel = CV.fit(train)

/databricks/spark/python/pyspark/ml/util.py:791: UserWarning: Can not find mlflow. To enable mlflow logging, install MLflow library from PyPi.
 warnings.warn(_MLflowInstrumentation._NO_MLFLOW_WARNING)

In [54]:
best_model = cvModel.bestModel
# best_model.rank
best_model.params

Out[58]: [Param(parent='ALS_eea967866f84', name='coldStartStrategy', doc='strategy for dealing with unknown or new users/items at prediction time. This may be useful in cross-validation or production scenarios, for handling user/item ids the model has not seen in the training data. Supported values: nan,drop.'),
 Param(parent='ALS_eea967866f84', name='itemCol', doc='column name for item ids. Ids must be within the integer value range.'),
 Param(parent='ALS_eea967866f84', name='predictionCol', doc='prediction column name'),
 Param(parent='ALS_eea967866f84', name='userCol', doc='column name for user ids. Ids must be within the integer value range.')]

In [55]:
cvmodel_path = '/FileStore/tables/movieLen' + '/best_cvModel'
best_model.save(cvmodel_path)

In [56]:
# cvmodel_path = '/FileStore/tables/movieLen' + '/best_cvModel'
cv_model = ALSModel.load(cvmodel_path)
cv_model.rank

Out[16]: 10

#### Method 2: use cross-evaluation toolbox

In [58]:
def train_ALS(train_data, validation_data, num_iters, reg_param, ranks):
    min_error = float('inf')
    best_rank = -1
    best_regularization = 0
    best_model = None
    for rank in ranks:
        for reg in reg_param:
            # write your approach to train ALS model
            # make prediction
            # get the rating result
            # get the RMSE
            als = ALS(rank=rank, maxIter=num_iters, regParam=reg, userCol="userId", itemCol="movieId", ratingCol="rating",
                      coldStartStrategy="drop")
            model = als.fit(train_data)
            predictions = model.transform(validation_data)
            
            evaluator = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
            error = evaluator.evaluate(predictions)
            print ('{} latent factors and regularization = {}: validation RMSE is {}'.format(rank, reg, error))
            if error < min_error:
                min_error = error
                best_rank = rank
                best_regularization = reg
                best_model = model
    print ('\nThe best model has {} latent factors and regularization = {}'.format(best_rank, best_regularization))
    return best_model, best_rank, best_regularization

In [59]:
num_iterations = 10
ranks = [6, 8, 10, 12, 14]
reg_params = [0.05, 0.1, 0.2, 0.4, 0.8]

import time
start_time = time.time()
final_model, final_rank, final_reg = train_ALS(training, validation, num_iterations, reg_params, ranks)

print ('Total Runtime: {:.2f} seconds'.format(time.time() - start_time))

6 latent factors and regularization = 0.05: validation RMSE is 0.8270339846917111
6 latent factors and regularization = 0.1: validation RMSE is 0.8259783723732318
6 latent factors and regularization = 0.2: validation RMSE is 0.865728671966565
6 latent factors and regularization = 0.4: validation RMSE is 0.9611198205310132
6 latent factors and regularization = 0.8: validation RMSE is 1.184876345583705
8 latent factors and regularization = 0.05: validation RMSE is 0.8225150965936567
8 latent factors and regularization = 0.1: validation RMSE is 0.8234703478902027
8 latent factors and regularization = 0.2: validation RMSE is 0.86647758131085
8 latent factors and regularization = 0.4: validation RMSE is 0.9606929629057956
8 latent factors and regularization = 0.8: validation RMSE is 1.1848832137336556
10 latent factors and regularization = 0.05: validation RMSE is 0.816098608015772
10 latent factors and regularization = 0.1: validation RMSE is 0.8175397923503945
10 latent factors and regularization = 0.2: validation RMSE is 0.8646292160443473
10 latent factors and regularization = 0.4: validation RMSE is 0.9624828164118169
10 latent factors and regularization = 0.8: validation RMSE is 1.1848536484547412
12 latent factors and regularization = 0.05: validation RMSE is 0.8187615936843773
12 latent factors and regularization = 0.1: validation RMSE is 0.8199634536925003
12 latent factors and regularization = 0.2: validation RMSE is 0.8658598059752847
12 latent factors and regularization = 0.4: validation RMSE is 0.9608932522795528
12 latent factors and regularization = 0.8: validation RMSE is 1.1848738607351403
14 latent factors and regularization = 0.05: validation RMSE is 0.8160381409737216
14 latent factors and regularization = 0.1: validation RMSE is 0.8185922754250311
14 latent factors and regularization = 0.2: validation RMSE is 0.8657723480591543
14 latent factors and regularization = 0.4: validation RMSE is 0.9611950712362183
14 latent factors and regularization = 0.8: validation RMSE is 1.1848696500682194

The best model has 14 latent factors and regularization = 0.05
Total Runtime: 4823.12 seconds

#### Plot learning rate

In [61]:
def learning_curve(iter_array, train, validation, rank, reg):
  train_errors, val_errors = [], []
  for iter in iter_array:
    als = ALS(rank=rank, maxIter=iter, regParam=reg, userCol="userId", itemCol="movieId", ratingCol="rating",
                      coldStartStrategy="drop")
    als_model = als.fit(train)
    train_predict = als_model.transform(train)
    val_predict = als_model.transform(validation)
    train_errors.append(mse_evaluator.evaluate(train_predict))
    val_errors.append(mse_evaluator.evaluate(val_predict))
  
  return train_errors, val_errors
    
    

In [62]:
iter_array = [1, 2, 5, 10]
train_errors, val_errors = learning_curve(iter_array, train, validation, final_rank, final_reg)

In [63]:
def plot_lr_curve(iter_array, train_errors, val_errors):
  # create dataframe
  lst = [iter_array, train_errors, val_errors]
  R = Row('iter_num', 'train_mse', 'validation_mse')
  lr_df = sc.parallelize(R(*r) for r in zip(*lst)).toDF()

  # plot learning curve
  fig, axes = plt.subplots(1,1)
  lr_df_p = lr_df.toPandas()
  plt.plot('iter_num','train_mse',data=lr_df_p)
  plt.plot('iter_num','validation_mse',data=lr_df_p)
  plt.xlabel("num_iteration")
  plt.ylabel("RMSE")
  plt.title("learning curve")
  plt.legend()

  display(fig.show())

In [64]:
plot_lr_curve(iter_array, train_errors, val_errors)

In [65]:
model_path = '/FileStore/tables/movieLen' + '/bestModel2'
final_model.save(model_path)

In [66]:
# model_path = '/FileStore/tables/movieLen' + '/bestModel2'
final_model = ALSModel.load(model_path)

### Model testing
And finally, make a prediction and check the testing error.

In [68]:
rmse1 = mse_evaluator.evaluate(cv_model.transform(test))
rmse2 = mse_evaluator.evaluate(final_model.transform(testing))
print('test rmse for method 1 is: {}'.format(rmse1))
print('test rmse for method 2 is: {}'.format(rmse2))

test rmse for method 1 is: 0.8159831533875902
test rmse for method 2 is: 0.816709903665061

### Apply the model to do recommendation

In [70]:
rec_to_user = final_model.recommendForAllUsers(10)

In [71]:
def get_recommend(recommend_to_user_df):
  recommend = (recommend_to_user_df
        .withColumn("tmp", arrays_zip("recommendations"))
        .withColumn("tmp", explode("tmp"))
        .select("userId", col("tmp.recommendations")['movieId'].alias('movieId'), col("tmp.recommendations")['rating'].alias('rating')))
  return recommend.join(movies, recommend.movieId==movies.movieId).select(recommend.userId, "title", "genres", "rating")

In [72]:
rec_to_user_df = get_recommend(rec_to_user)
display(rec_to_user_df)

userId title genres rating 148 Heroes (2008) (no genres listed) 5.7512302 148 The War at Home (1979) Documentary|War 5.6723747 148 Glenn Killing på Berns (1993) Comedy 5.396998 148 The Great Piggy Bank Robbery (1946) Animation|Children|Comedy 5.3768034 148 Rising Place, The (2002) Drama 5.242195 148 The Old Gun (1975) Drama|Thriller|War 5.1790867 148 Ali Baba Bunny (1957) Animation|Children|Comedy 5.108641 148 Sharpe's Eagle (1993) Action|Adventure|War 5.0117536 148 The Yellow Rolls-Royce (1964) Comedy|Drama|Romance 5.00092 148 Freedom on My Mind (1994) Documentary 4.9937572 463 I Can Speak (2017) Comedy|Drama 3.3854594 463 Fifi Howls from Happiness (2013) Documentary 3.2775798 463 Acéphale (1969) (no genres listed) 3.2685142 463 Rammstein: Live aus Berlin (1999) (no genres listed) 3.2612965 463 They Call Us Misfits (Dom kallar oss mods) (1968) Documentary 3.2527223 463 Morgan Murphy: Irish Goodbye (2014) Comedy 3.2434483 463 Holy Innocents, The (Santos inocentes, Los) (1984) Drama 3.2398853 463 Freedom on My Mind (1994) Documentary 3.2227457 463 Heroes (2008) (no genres listed) 3.2202609 463 Judah Friedlander: America Is the Greatest Country in the United States (2017) Comedy 3.219542 471 Blood Brother (2013) Documentary 4.731188 471 Unfair Competition (Concorrenza sleale) (2001) Drama|War 4.678417 471 The Winslow Boy (1948) Children|Crime|Drama 4.58308 471 Seeing Red: Stories of American Communists (1983) (no genres listed) 4.572272 471 Day the Universe Changed, The (1985) Documentary 4.526875 471 Freedom on My Mind (1994) Documentary 4.5063453 471 Nazis: A Warning from History, The (1997) Documentary|War 4.5051403 471 Doggiewoggiez! Poochiewoochiez! (2012) Comedy 4.4693565 471 Tito and Me (Tito i ja) (1992) Comedy 4.4628053 471 NOFX Backstage Passport 2 (no genres listed) 4.4595733 496 Eve and the Fire Horse (2005) Drama 6.4870315 496 UFO - Distruggete base Luna! (1971) Action|Sci-Fi|Thriller 6.330229 496 Horror Stories (2012) Horror|Thriller 5.812172 496 Ward 13 (2003) Action|Animation|Comedy|Horror 5.7733426 496 The Spirit of Christmas (1995) Animation|Comedy 5.634837 496 14+ (2015) Drama|Romance 5.623262 496 Femmes femmes (1974) Comedy|Drama 5.4831457 496 Ashik Kerib (1988) Drama|Romance 5.474144 496 Don't Laugh at My Romance (2008) Comedy|Drama 5.4562497 496 Who the Hell Is Juliette? (¿Quién diablos es Juliette?) (1997) Documentary 5.4469614 833 Everybody in Our Family (2012) Drama 7.3890004 833 Be My Cat: A Film for Anne (2015) Horror|Thriller 6.8615885 833 Cat City (1986) Animation|Children 6.843827 833 Papers, Please: The Short Film (2018) Drama|Thriller 6.704534 833 District, The (Nyócker!) (2004) Animation|Comedy 6.3677254 833 The Legend of Paul and Paula (1973) Drama|Romance 6.347969 833 Mudhoney (1965) Drama 6.3043404 833 Blade, The (Dao) (1995) Action|Drama 6.251496 833 UFO - Distruggete base Luna! (1971) Action|Sci-Fi|Thriller 6.153562 833 Big Snit, The (1985) Animation|Comedy 6.1408844 1088 Heroes (2008) (no genres listed) 5.522358 1088 UFO - Distruggete base Luna! (1971) Action|Sci-Fi|Thriller 4.9837575 1088 Ethan Mao (2004) Drama|Thriller 4.951622 1088 Dumb: The Story of Big Brother Magazine (2017) Documentary 4.798112 1088 Last Lions, The (2011) Documentary 4.7746015 1088 Fantastic Lies (2016) Documentary 4.6641626 1088 Everybody in Our Family (2012) Drama 4.614707 1088 Demetri Martin. Person. (2007) Comedy 4.581496 1088 You're Gonna Miss Me (2005) Documentary 4.5330296 1088 The Legend of Paul and Paula (1973) Drama|Romance 4.512246 1238 Peppermint Soda (Diabolo menthe) (1977) Comedy|Drama 3.614352 1238 Blood Brother (2013) Documentary 3.59673 1238 Murder, She Baked: A Plum Pudding Murder Mystery (2015) Mystery 3.5902405 1238 Plan B (2009) Comedy|Drama|Romance 3.5462554 1238 Queen: Days of Our Lives (2011) Documentary 3.5229576 1238 Moliere (1978) Drama 3.472147 1238 Shiloh 2: Shiloh Season (1999) Children|Drama 3.3859348 1238 The Wilde Wedding (2017) Comedy 3.3769019 1238 Mission London (2010) Comedy 3.371

In [73]:
display(rec_to_user_df.describe())

summary userId title genres rating count 2798690 2798690 2798690 2798690 mean 141630.38205017347 null null 5.646920058055335 stddev 81767.65366554947 null null 0.8499038642426338 min 1 """Great Performances"" Cats (1998)" (no genres listed) 0.45407003 max 283228 チェブラーシカ (2010) Western 13.582502

#### Personalized recommendation

In [75]:
user_subset = ratings.where(ratings.userId == 1)
user_subset_recs = get_recommend(final_model.recommendForUserSubset(user_subset, 10))


In [76]:
display(user_subset_recs)

userId title genres rating 1 Willie & Phil (1980) Comedy|Drama|Romance 5.9306693 1 O Pátio das Cantigas (1942) Comedy 5.832705 1 Let's Play Two (2017) Documentary 5.802419 1 The State I Am In (2000) Drama 5.708998 1 Pearl Jam: Immagine in Cornice - Live in Italy 2006 (2007) Documentary|Musical 5.60396 1 What Did You Do in the War, Thanassi? (1971) Comedy|Drama 5.575147 1 Ο Θανάσης στη χώρα της σφαλιάρας (1976) (no genres listed) 5.5695586 1 Garbage Warrior (2007) Documentary 5.5587406 1 Return to Source: The Philosophy of The Matrix (2004) Documentary 5.5492215 1 Day of the Wacko (Dzien swira) (2002) Comedy|Drama 5.5443153

In [77]:
# recommend to new users

# 1. Adding the new ratings into the DataFrame (hint: look into using the .union() method)
# 2. Fitting the ALS model
# 3. Make recommendations for the user of choice
# 4. Print out the names of the top $n$ recommendations in a reader-friendly manner


# self_def_user = sc.parallelize([(7654321, 1, 4.0), (7654321, 16, 3.5), (7654321, 40, 4.5), (7654321, 78, 3.5), (7654321, 90, 2.0), 
#                                 (7654321, 100, 1.0), (7654321, 200, 5.0), (7654321, 352, 4.0), (7654321, 508, 4.5), (7654321, 1200, 3.5)]).toDF(("userId", "movieId", "rating"))

# rec_to_self_def_user = get_recommend(final_model.recommendForUserSubset(self_def_user, 2))

